devito/ir/clusters/algorithms.py

```diff
@@ -368,19 +369,46 @@ def pull_indexeds(expr, subs, mapper, parent=None):
 
 
 @cluster_pass(mode='all')
-def normalize_reductions(cluster, sregistry):
+def normalize_reductions(cluster, sregistry, options):
     """
     Extract the right-hand sides of reduction Eq's in to temporaries.
     """
-    if not any(PARALLEL_IF_ATOMIC in v for v in cluster.properties.values()):
+    opt_mapify_reduce = options['mapify-reduce']
+
+    dims = [d for d, v in cluster.properties.items() if PARALLEL_IF_ATOMIC in v]
+
+    if not dims:
         return cluster
 
     processed = []
     for e in cluster.exprs:
-        if e.is_Increment and e.lhs.function.is_AbstractFunction:
-            v = Symbol(name=sregistry.make_name(), dtype=e.dtype)
+        if e.is_Reduction and e.lhs.is_Indexed and cluster.is_sparse:
+            # Transform `e` such that we reduce into a scalar (ultimately via
+            # atomic ops, though this part is carried out by a much later pass)
+            # For example, given `i = m[p_src]` (i.e., indirection array), turn:
+            # `u[t, i] += f(u[t, i], src, ...)`
+            # into
+            # `s = f(u[t, i], src, ...)`
+            # `u[t, i] += s`
+            name = sregistry.make_name()
+            v = Symbol(name=name, dtype=e.dtype)
             processed.extend([e.func(v, e.rhs, operation=None),
                               e.func(e.lhs, v)])
+
+        elif e.is_Reduction and e.lhs.is_Symbol and opt_mapify_reduce:
+            # Transform `e` into what is in essence an explicit map-reduce
+            # For example, turn:
+            # `s += f(u[x], v[x], ...)`
+            # into
+            # `r[x] = f(u[x], v[x], ...)`
```

Comment:
Is this a block size temp or full size temp?

Reply:
Can this be turned off with `min-storage` ?

Reply:
full size temporary

> Can this be turned off with `min-storage`

You're overthinking it probably. This is by default disabled because `opt_mapify_reduce` is set to False in all backends

Reply:
turning it into into a block-size temporary would require quite a bit of work, but I agree in principle it should be like you say